capstone/capdb/models.py

```diff
@@ -842,12 +856,169 @@ def create_or_update_case_text(self, new_text=None):
             case_text = CaseText(metadata=self)
 
         if new_text is None:
-            new_text = self.case_xml.extract_casebody().text()
+            try:
+                new_text = self.structure.extract_text()
+            except CaseStructure.DoesNotExist:
+                new_text = self.case_xml.extract_casebody().text()
 
         if new_text != case_text.text:
             case_text.text = new_text
             case_text.save()
 
+    def get_hydrated_structure(self):
+        """
+            Return structure for this case with all references to page blocks replaced by the actual block contents.
+            This is useful for diagnosing rendering errors.
+        """
+        structure = self.structure
+        blocks_by_id = PageStructure.blocks_by_id(structure.pages.all())
+        renderer = render_case.VolumeRenderer(blocks_by_id, {}, {})
+        return renderer.hydrate_opinions(structure.opinions, blocks_by_id)
+
+    def sync_case_body_cache(self, blocks_by_id=None, fonts_by_id=None, labels_by_block_id=None):
+        """
+            Update self.body_cache with new values based on the current value of self.structure.
+            blocks_by_id and fonts_by_id can be provided for efficiency if updating a bunch of cases from the same volume.
+        """
+        structure = self.structure
+        if not blocks_by_id or not labels_by_block_id:
+            pages = list(structure.pages.all())
+        blocks_by_id = blocks_by_id or PageStructure.blocks_by_id(pages)
+        fonts_by_id = fonts_by_id or CaseFont.fonts_by_id(blocks_by_id)
+        labels_by_block_id = labels_by_block_id or PageStructure.labels_by_block_id(pages)
+
+        renderer = render_case.VolumeRenderer(blocks_by_id, fonts_by_id, labels_by_block_id)
+        text = renderer.render_text(self)
+        html = renderer.render_html(self)
+        xml = renderer.render_xml(self)
+
+        ## create json
+
+        casebody_pq = PyQuery(html)
+
+        # add footnote labels to footnotes
+        for footnote in casebody_pq('.footnote').items():
+            label = footnote.attr['data-label']
+            if label:
+                footnote_paragraph = footnote.children()[0]
+                footnote_paragraph.text = "%s %s" % (label, footnote_paragraph.text)
```

Comment:
Do we ever do anything with `footnote_paragraph`? I can't see where.

Reply:
This modifies the value of casebody_pq, which then shows up in opinion.text() down below.